rudder-core/src/main/scala/com/normation/rudder/services/reports/ExecutionBatch.scala

```diff
@@ -583,15 +645,43 @@ object ExecutionBatch extends Loggable {
                                    missing ++ unexpected ++ expected
                                 }
     } yield {
-      RuleNodeStatusReport(
-          mergeInfo.nodeId
-        , ruleId
-        , serial
-        , mergeInfo.run
-        , mergeInfo.configId
-        , DirectiveStatusReport.merge(directiveStatusReports)
+      (
+          SerialedRuleId(ruleId, serial)
+        , RuleNodeStatusReport(
+              mergeInfo.nodeId
+            , ruleId
+            , serial //can not be empty because of groupBy
+            , mergeInfo.run
+            , mergeInfo.configId
+            , DirectiveStatusReport.merge(directiveStatusReports)
+            , mergeInfo.expirationTime
+          )
       )
-    }).toSeq
+    }).toMap
+
+    //now, for all current expected reports, choose between the computed value and the default one
+
+    // note: isn't there something specific to do for unexpected reports ? Keep them all ?
+
+    val nil = Seq[RuleNodeStatusReport]()
+    val (computed, newStatus) = ((nil, nil)/: currentExpectedDirectives) { case ( (c,n), (k, expectedReport)) =>
+      complianceForRun.get(k) match {
+        case None => //the whole rule is new!
+          //here, the reports are ACTUALLY pending, not missing.
+          val x = buildRuleNodeStatusReport(mergeInfo, Map(k -> expectedReport), PendingReportType)
```

Comment:
i'm not sure i understand here the k, and why the reports are pendings
But does that mean that the pending state will never expire is we don't receive reports ?